File: client/src/main/js/components/CourseCreationModal.js
import React from 'react';
import {Api} from '../Api';
import Immutable from 'immutable';
import {credentials} from '../scripts';
import ReactDOM from 'react-dom';
import {Notification} from './Notification';
import {
    Button,
    Form,
    FormGroup,
    FormText,
    Input,
    Label,
    Modal,
    ModalBody,
    ModalFooter,
    ModalHeader
} from 'reactstrap';

export class CourseCreationModal extends React.Component {

    constructor(props) {
        super(props);

        this.getLanguages = this.getLanguages.bind(this);
        this.getTestLanguages = this.getTestLanguages.bind(this);
        this.getTestingFrameworks = this.getTestingFrameworks.bind(this);
        this.toggle = this.toggle.bind(this);
        this.handleCourseNameChange = this.handleCourseNameChange.bind(this);
        this.handleCourseDescriptionChange = this.handleCourseDescriptionChange.bind(this);
        this.handleLanguageChange = this.handleLanguageChange.bind(this);
        this.handleTestingLanguageChange = this.handleTestingLanguageChange.bind(this);
        this.handleTestingFrameworkChange = this.handleTestingFrameworkChange.bind(this);
        this.handleNumberOfTasksChange = this.handleNumberOfTasksChange.bind(this);
        this.createCourse = this.createCourse.bind(this);

        this.state = {
            show: false,
            courseName: null,
            courseDescription: null,
            language: null,
            testLanguage: null,
            testingFramework: null,
            numberOfTasks: null,
            flaxoLanguages: []
        };

        Api.retrieveLanguages(
            languages => {
                this.setState(
                    {
                        flaxoLanguages: languages,
                        language: languages[0].name,
                        testLanguage: languages[0].compatibleTestingLanguages[0],
                        testingFramework: languages[0].compatibleTestingFrameworks[0]
                    }
                )
            }
        );
    }

    toggle() {
        this.setState({show: !this.state.show});
    }

    getLanguages() {
        return (
            Immutable.List(this.state.flaxoLanguages)
                .map(language => language.name)
                .map(name => <option value={name}
                                     selected={name === this.state.language ? "selected" : ""}>{name}</option>
                )
        );
    }

    getTestLanguages() {
        return (
            Immutable.List(this.state.flaxoLanguages)
                .filter(language => language.name === this.state.language)
                .flatMap(language => language.compatibleTestingLanguages)
                .map(name => <option value={name}
                                     selected={name === this.state.testLanguage ? "selected" : ""}>{name}</option>
                )
        );
    }

    getTestingFrameworks() {
        return (
            Immutable.List(this.state.flaxoLanguages)
                .filter(language => language.name === this.state.testLanguage)
                .flatMap(language => language.compatibleTestingFrameworks)
                .map(name => <option value={name}
                                     selected={name === this.state.testingFramework ? "selected" : ""}>{name}</option>
                )
        );
    }

    handleCourseNameChange(event) {
        this.setState({courseName: event.target.value});
    }

    handleCourseDescriptionChange(event) {
        this.setState({courseDescription: event.target.value});
    }

    handleLanguageChange(event) {
        const selectedLanguage = event.target.value;

        const defaultTestLanguage =
            Immutable.List(this.state.flaxoLanguages)
                .find(language => language.name === selectedLanguage)
                .compatibleTestingLanguages[0];

        const defaultTestingFramework =
            Immutable.List(this.state.flaxoLanguages)
                .find(language => language.name === defaultTestLanguage)
                .compatibleTestingFrameworks[0];

        this.setState({
            language: selectedLanguage,
            testLanguage: defaultTestLanguage,
            testingFramework: defaultTestingFramework
        });
    }

    handleTestingLanguageChange(event) {
        const selectedTestLanguage = event.target.value;

        const defaultTestingFramework =
            Immutable.List(this.state.flaxoLanguages)
                .find(language => language.name === selectedTestLanguage)
                .compatibleTestingFrameworks[0];

        this.setState({
            testLanguage: selectedTestLanguage,
            testingFramework: defaultTestingFramework
        });
    }

    handleTestingFrameworkChange(event) {
        this.setState({testingFramework: event.target.value});
    }

    handleNumberOfTasksChange(event) {
        this.setState({numberOfTasks: event.target.value});
    }

    createCourse() {
        Api.createCourse(credentials(),
            {
                courseName: this.state.courseName,
                description: this.state.courseDescription,
                language: this.state.language,
                testingLanguage: this.state.testLanguage,
                testingFramework: this.state.testingFramework,
                numberOfTasks: parseInt(this.state.numberOfTasks)
            },
            () => {
                ReactDOM.render(
                    <Notification succeed
                                  message={`Course ${this.state.courseName} has been created`}/>,
                    document.getElementById('notifications')
                );

                this.props.onCourseCreation();
            },
            response => ReactDOM.render(
                <Notification message={`Course ${this.state.courseName} creation failed due to: ${response}`}/>,
                document.getElementById('notifications')
            )
        );

        this.toggle();
    }

    render() {
        return (
            <div>
                <Button color="primary" outline
                        onClick={this.toggle}>Create course</Button>
                <Modal isOpen={this.state.show} toggle={this.toggle}>
                    <ModalHeader toggle={this.toggle}>Create course</ModalHeader>
                    <ModalBody>
                        <Form>
                            <FormGroup>
                                <Label for="courseName">Course name</Label>
                                <Input type="text" id="courseName" onChange={this.handleCourseNameChange}/>
                                <FormText color="muted">Course name should be a valid git repository name</FormText>
                            </FormGroup>
                            <FormGroup>
                                <Label for="courseDescription">Course description <small className="text-muted">optional</small></Label>
                                <Input type="text" id="courseDescription" onChange={this.handleCourseDescriptionChange}/>
                                <FormText color="muted">Course description won't be visible for students</FormText>
                            </FormGroup>
                            <FormGroup>
                                <Label for="language">Language</Label>
                                <Input type="select" id="language" onChange={this.handleLanguageChange}>
                                    {this.getLanguages()}
                                </Input>
                                <FormText color="muted">Language solutions will be written on</FormText>
                            </FormGroup>
                            <FormGroup>
                                <Label for="testingLanguage">Testing language</Label>
                                <Input type="select" id="testingLanguage" onChange={this.handleTestingLanguageChange}>
                                    {this.getTestLanguages()}
                                </Input>
                                <FormText color="muted">Language tests will be written on</FormText>
                            </FormGroup>
                            <FormGroup>
                                <Label for="testingFramework">Testing framework</Label>
                                <Input type="select" id="testingFramework" onChange={this.handleTestingFrameworkChange}>
                                    {this.getTestingFrameworks()}
                                </Input>
                                <FormText color="muted">Test framework to use in course</FormText>
                            </FormGroup>
                            <FormGroup>
                                <Label for="testingFramework">Number of tasks</Label>
                                <Input type="text" id="testingFramework" onChange={this.handleNumberOfTasksChange}/>
                                <FormText color="muted">Number of git branches for tasks</FormText>
                            </FormGroup>
                        </Form>
                    </ModalBody>
                    <ModalFooter>
                        <Button color="primary" onClick={this.createCourse}>Create</Button>{' '}
                        <Button color="secondary" onClick={this.toggle}>Cancel</Button>
                    </ModalFooter>
                </Modal>
            </div>
        );
    }

}

File: frontend/src/main/kotlin/org/flaxo/frontend/component/CourseCreationModal.kt
package org.flaxo.frontend.component

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.html.ButtonType
import kotlinx.html.InputType
import kotlinx.html.classes
import kotlinx.html.hidden
import kotlinx.html.id
import kotlinx.html.js.onClickFunction
import kotlinx.html.role
import kotlinx.html.tabIndex
import org.flaxo.common.Framework
import org.flaxo.common.Language
import org.flaxo.frontend.Container
import org.flaxo.frontend.Notifications
import org.flaxo.frontend.OnCourseChange
import org.flaxo.frontend.checkBoxValue
import org.flaxo.frontend.clickOnButton
import org.flaxo.frontend.client.FlaxoClient
import org.flaxo.frontend.client.FlaxoHttpException
import org.flaxo.frontend.component.bootstrap.selectComponent
import org.flaxo.frontend.credentials
import org.flaxo.frontend.inputValue
import org.flaxo.frontend.selectValue
import org.flaxo.frontend.validatedInputValue
import react.RBuilder
import react.RComponent
import react.RProps
import react.RState
import react.dom.b
import react.dom.button
import react.dom.div
import react.dom.h5
import react.dom.input
import react.dom.small
import react.dom.span
import react.setState

const val COURSE_CREATION_MODAL_ID = "courseCreationModal"
private const val COURSE_NAME_INPUT_ID = "courseNameInput"
private const val COURSE_NAME_INPUT_HELP_ID = "courseNameInputHelp"
private const val COURSE_DESCRIPTION_INPUT_ID = "courseDescriptionInput"
private const val COURSE_DESCRIPTION_INPUT_HELP_ID = "courseDescriptionInputHelp"
private const val PRIVATE_COURSE_CHECKBOX_ID = "privateCourseCheckbox"
private const val PRIVATE_COURSE_CHECKBOX_HELP_ID = "privateCourseCheckboxHelp"
private const val GENERATE_ENVIRONMENT_CHECKBOX_ID = "generateEnvironmentCheckbox"
private const val GENERATE_ENVIRONMENT_CHECKBOX_HELP_ID = "generateEnvironmentCheckboxHelp"
private const val LANGUAGE_SELECT_ID = "languageSelect"
private const val TESTING_LANGUAGE_SELECT_ID = "testingLanguageSelect"
private const val TESTING_FRAMEWORK_SELECT_ID = "testingFrameworkSelect"
private const val NUMBER_OF_TASKS_INPUT_ID = "numberOfTasksInput"
private const val NUMBER_OF_TASKS_INPUT_HELP_ID = "numberOfTasksInputHelp"
private const val COURSE_CREATION_MODAL_CANCEL_ID = "courseCreationModalCancel"

/**
 * Adds course creation modal.
 */
fun RBuilder.courseCreationModal(onCreate: OnCourseChange) = child(CourseCreationModal::class) {
    attrs {
        this.onCreate = onCreate
    }
}

private class CourseCreationModalProps(var onCreate: OnCourseChange) : RProps

private class CourseCreationModalState(var private: Boolean = false,
                                       var generateEnvironment: Boolean = false,
                                       var language: Language? = null,
                                       var testingLanguage: Language? = null,
                                       var testingFramework: Framework? = null
) : RState

private class CourseCreationModal(props: CourseCreationModalProps)
    : RComponent<CourseCreationModalProps, CourseCreationModalState>(props) {

    private val flaxoClient: FlaxoClient

    init {
        state = CourseCreationModalState(language = Language.Java, testingLanguage = Language.Java,
                testingFramework = Framework.JUnit)
        flaxoClient = Container.flaxoClient
    }

    override fun RBuilder.render() {
        div("modal fade") {
            attrs {
                id = COURSE_CREATION_MODAL_ID
                tabIndex = "-1"
                role = "dialog"
                ariaHidden = true
            }
            div("modal-dialog") {
                attrs {
                    role = "document"
                }
                div("modal-content") {
                    div("modal-header") {
                        h5("modal-title") {
                            +"Create course"
                        }
                        button(classes = "close", type = ButtonType.button) {
                            attrs {
                                dataDismiss = "modal"
                                ariaLabel = "Close"
                            }
                            span {
                                attrs {
                                    hidden = true
                                }
                                +"&times;"
                            }
                        }
                    }
                    div("modal-body") {
                        courseNameInput()
                        courseDescriptionInput()
                        privateCourseCheckbox()
                        tasksNumberInput()
                        languageSelect()
                        generateEnvironmentCheckbox()
                        if (state.generateEnvironment) {
                            testingLanguageSelect()
                            testingFrameworkSelect()
                        }
                    }
                    div("modal-footer") {
                        button(classes = "btn btn-primary", type = ButtonType.button) {
                            attrs {
                                onClickFunction = { GlobalScope.launch { createCourse() } }
                            }
                            +"Create"
                        }
                        button(classes = "btn btn-secondary", type = ButtonType.button) {
                            attrs {
                                id = COURSE_CREATION_MODAL_CANCEL_ID
                                dataDismiss = "modal"
                            }
                            +"Cancel"
                        }
                    }
                }
            }
        }
    }

    private fun RBuilder.courseNameInput() {
        div("form-group") {
            label("Course name", COURSE_NAME_INPUT_ID)
            input {
                attrs {
                    id = COURSE_NAME_INPUT_ID
                    classes = setOf("form-control")
                    type = InputType.text
                    ariaDescribedBy = COURSE_NAME_INPUT_HELP_ID
                }
            }
            small {
                attrs {
                    id = COURSE_NAME_INPUT_HELP_ID
                    classes = setOf("form-text", "text-muted")
                }
                +"Course name should be a valid git repository name"
            }
        }
    }

    private fun RBuilder.courseDescriptionInput() {
        div("form-group") {
            label("Course description", COURSE_DESCRIPTION_INPUT_ID)
            input {
                attrs {
                    id = COURSE_DESCRIPTION_INPUT_ID
                    classes = setOf("form-control")
                    type = InputType.text
                    ariaDescribedBy = COURSE_DESCRIPTION_INPUT_HELP_ID
                }
            }
            small {
                attrs {
                    id = COURSE_DESCRIPTION_INPUT_HELP_ID
                    classes = setOf("form-text", "text-muted")
                }
                +"Course description won't be visible for students"
            }
        }
    }

    private fun RBuilder.privateCourseCheckbox() {
        div("form-group") {
            div("form-check") {
                input {
                    attrs {
                        id = PRIVATE_COURSE_CHECKBOX_ID
                        classes = setOf("form-check-input")
                        type = InputType.checkBox
                        ariaDescribedBy = PRIVATE_COURSE_CHECKBOX_HELP_ID
                        defaultChecked = false
                        onClickFunction = { setState { private = !private } }
                    }
                }
                label("Private course", PRIVATE_COURSE_CHECKBOX_ID, classes = setOf("checkbox-label"))
            }
            small {
                attrs {
                    id = PRIVATE_COURSE_CHECKBOX_HELP_ID
                    classes = setOf("form-text", "text-muted", "checkbox-help")
                }
                +"Specifies either the creating course should be private or public (default)."
            }
        }
    }


    private fun RBuilder.generateEnvironmentCheckbox() {
        val isDisabled = state.language?.testingLanguages.isNullOrEmpty()
        div("form-group") {
            div("form-check") {
                input {
                    attrs {
                        id = GENERATE_ENVIRONMENT_CHECKBOX_ID
                        classes = setOf("form-check-input")
                        type = InputType.checkBox
                        ariaDescribedBy = GENERATE_ENVIRONMENT_CHECKBOX_HELP_ID
                        defaultChecked = state.generateEnvironment
                        disabled = isDisabled
                        onClickFunction = { setState { generateEnvironment = !generateEnvironment } }
                    }
                }
                val checkboxLabels =
                        if (isDisabled) setOf("checkbox-label", "text-muted")
                        else setOf("checkbox-label")
                label("Generate environment", GENERATE_ENVIRONMENT_CHECKBOX_ID, classes = checkboxLabels)
            }
            small {
                attrs {
                    id = GENERATE_ENVIRONMENT_CHECKBOX_HELP_ID
                    classes = setOf("form-text", "text-muted", "checkbox-help")
                }
                +"Environment generation is an "
                b { +"experiment feature " }
                +"that is available only for several languages and frameworks."
            }
        }
    }

    private fun RBuilder.languageSelect(): Unit = selectComponent(
            selectId = LANGUAGE_SELECT_ID,
            name = "Language",
            description = "Programming language that will be used by the course students in their solutions. "
                    + "It should be specified in order to perform plagiarism analysis.",
            default = state.language?.alias,
            options = Language.values()
                    .filter { it.testingLanguages.isNotEmpty() || !state.generateEnvironment }
                    .map { it.alias },
            onUpdate = {
                setState {
                    language = Language.from(it)
                    testingLanguage = language?.testingLanguages.orEmpty()
                            .firstOrNull { it.testingFrameworks.isNotEmpty() }
                    testingFramework = testingLanguage?.testingFrameworks?.firstOrNull()
                }
            }
    )

    private fun RBuilder.testingLanguageSelect(): Unit = selectComponent(
            selectId = TESTING_LANGUAGE_SELECT_ID,
            name = "Testing language",
            description = "Programming language that will be used by the course author in task specifications. "
                    + "It should be specified in order to autobuild project infrastructure.",
            default = state.testingLanguage?.alias,
            options = state.language?.testingLanguages.orEmpty().map { it.alias },
            onUpdate = {
                setState {
                    testingLanguage = Language.from(it)
                    testingFramework = testingLanguage?.testingFrameworks?.firstOrNull()
                }
            }
    )

    private fun RBuilder.testingFrameworkSelect() = selectComponent(
            selectId = TESTING_FRAMEWORK_SELECT_ID,
            name = "Testing framework",
            description = "Testing framework that will be used with the corresponding testing language " +
                    "by the course author in task specifications. It should be specified in order to autobuild an " +
                    "associated repository infrastructure.",
            default = state.testingFramework?.alias,
            options = state.testingLanguage?.testingFrameworks.orEmpty().map { it.alias }
    )

    private fun RBuilder.tasksNumberInput() {
        div("form-group") {
            label("Number of tasks", NUMBER_OF_TASKS_INPUT_ID)
            input {
                attrs {
                    id = NUMBER_OF_TASKS_INPUT_ID
                    classes = setOf("form-control")
                    type = InputType.number
                    ariaDescribedBy = NUMBER_OF_TASKS_INPUT_HELP_ID
                }
            }
            small {
                attrs {
                    id = NUMBER_OF_TASKS_INPUT_HELP_ID
                    classes = setOf("form-text", "text-muted")
                }
                +"Number of tasks in course. Each branch represents a single task"
            }
        }
    }

    private suspend fun createCourse() {
        credentials?.also {
            try {
                val courseName = validatedInputValue(COURSE_NAME_INPUT_ID)
                val description = inputValue(COURSE_NAME_INPUT_ID)
                val private = checkBoxValue(PRIVATE_COURSE_CHECKBOX_ID)
                val generateEnvironment = checkBoxValue(GENERATE_ENVIRONMENT_CHECKBOX_ID)
                val numberOfTasks = validatedInputValue(NUMBER_OF_TASKS_INPUT_ID)?.toInt()
                val language = selectValue(LANGUAGE_SELECT_ID)
                val testingLanguage = selectValue(TESTING_LANGUAGE_SELECT_ID)
                val testingFramework = selectValue(TESTING_FRAMEWORK_SELECT_ID)
                if (courseName != null && numberOfTasks != null) {
                    clickOnButton(COURSE_CREATION_MODAL_CANCEL_ID)
                    Notifications.info("Course creation has been started.")
                    val createdCourse = flaxoClient.createCourse(it,
                            courseName = courseName,
                            description = description,
                            private = private,
                            language = language,
                            testingLanguage = testingLanguage?.filter { generateEnvironment },
                            testingFramework = testingFramework?.filter { generateEnvironment },
                            numberOfTasks = numberOfTasks)
                    props.onCreate(createdCourse)
                    Notifications.success("Course has been created.")
                }
            } catch (e: FlaxoHttpException) {
                console.log(e)
                Notifications.error("Error occurred while course creation.", e)
            }
        }
    }

}
